File: cmd/battery.ts
import { exec } from "./utils";

export default function battery(id: string) {
    const data = exec(`adb -s ${id} shell dumpsys battery`)
    if (data.error) throw new Error(data.stderr)
    const status_temp = +data.stdout.match(/status:\s*(\d)/)![1]
    let status: "charging" | "discharging" | "not charging" | "full" | "unknown"
    switch (status_temp) {
        // case 1: status = "unknown"; break;
        case 2: status = "charging"; break;
        case 3: status = "discharging"; break;
        case 4: status = "not charging"; break;
        case 5: status = "full"; break;
        default: status = "unknown";
    }
    const health_temp = +data.stdout.match(/health:\s*(\d)/)![1]
    let health: "unknown" | "good" | "overheat" | "dead" | "over voltage" | "unspecified failure" | "cold" | "unknown error"
    switch (health_temp) {
        // case 1: status = "unknown"; break;
        case 2: health = "good"; break;
        case 3: health = "overheat"; break;
        case 4: health = "dead"; break;
        case 5: health = "over voltage"; break;
        case 6: health = "unspecified failure"; break;
        case 7: health = "cold"; break;
        case 8: health = "unknown error"; break;
        default: health = "unknown";
    }
    const temperature = +data.stdout.match(/temperature:\s*(\d+)/)![1] / 10
    const technology = data.stdout.match(/technology:\s*(.+)/)![1]
    const voltage = +data.stdout.match(/voltage:\s*(\d+)/)![1] / 1000
    const level = +data.stdout.match(/level:\s*(\d+)/)![1]
    return { status, health, temperature, technology, voltage, level }
}

File: go/cli/battery.go
package cli

import (
	"fmt"
	"regexp"
	"strconv"
	"roidy/utils"
)

func getStatus(data string) string {
	// match status
	status_pattern := regexp.MustCompile(`status:\s*(\d)`)
	status_temp := status_pattern.FindStringSubmatch(data)
	// decide status
	switch status_temp[1] {
	case "2":
		return "charging"
	case "3":
		return "discharging"
	case "4":
		return "not_charging"
	case "5":
		return "full"
	default:
		return "unknown"
	}
}

func getHealth(data string) string {
	health_pattern := regexp.MustCompile(`health:\s*(\d)`)
	health_temp := health_pattern.FindStringSubmatch(data)
	switch health_temp[1] {
	case "2":
		return "good"
	case "3":
		return "overheat"
	case "4":
		return "dead"
	case "5":
		return "over voltage"
	case "6":
		return "unspecifed failure"
	default:
		return "unknown error"
	}
}

type BatteryInfo struct {
	status      string
	health      string
	temperature float64
	technology  string
	voltage     float64
	level       uint64
}

// fetch battery info of a device
func Battery(id string) (BatteryInfo, error) {
	data, err := utils.Execs(fmt.Sprintf("adb -s %s shell dumpsys battery", id))
	if err != nil {
		// panic(err)
		return BatteryInfo{}, err
	}
	status := getStatus(data)
	health := getHealth(data)
	temp_temp := regexp.MustCompile(`temperature:\s*(\d+)`).FindStringSubmatch(data)[1]
	temp_value, err := strconv.ParseFloat(temp_temp, 64)
	if err != nil {
		// panic(err)
		return BatteryInfo{}, err
	}
	temperature := temp_value / 10
	technology := regexp.MustCompile(`technology:\s*(.+)`).FindStringSubmatch(data)[1]
	volt_temp := regexp.MustCompile(`voltage:\s*(\d+)`).FindStringSubmatch(data)[1]
	volt_value, err := strconv.ParseFloat(volt_temp, 64)
	if err != nil {
		// panic(err)
		return BatteryInfo{}, err
	}
	voltage := volt_value / 1000
	level_temp := regexp.MustCompile(`level:\s*(\d+)`).FindStringSubmatch(data)[1]
	level, err := strconv.ParseUint(level_temp, 10, 8)
	if err != nil {
		// panic(err)
		return BatteryInfo{}, err
	}
	return BatteryInfo{
		status,
		health,
		temperature,
		technology,
		voltage,
		level,
	}, nil
}
